### ISSUE TITLE
The OpenTelemetry collector failed to export data because the ingester became unavailable.

### ISSUE THREAD

**Issue by edgarkz** [grafana/tempo#4493]
**Describe the bug**
Hello,
Tempo ingester shutdown is causing trace spans to be dropped due to failed otel collector export operations.

```
2024-12-25T03:58:10.168Z	error	exporterhelper/queue_sender.go:97	Exporting failed. Dropping data.	{"kind": "exporter", "data_type": "traces", "name": "otlp", "error": "not retryable error: Permanent error: rpc error: code = Unknown desc = Ingester is shutting down", "dropped_items": 1483}
go.opentelemetry.io/collector/exporter/exporterhelper.newQueueSender.func1
	go.opentelemetry.io/collector/exporter@v0.94.1/exporterhelper/queue_sender.go:97
go.opentelemetry.io/collector/exporter/internal/queue.(*boundedMemoryQueue[...]).Consume
	go.opentelemetry.io/collector/exporter@v0.94.1/internal/queue/bounded_memory_queue.go:57
go.opentelemetry.io/collector/exporter/internal/queue.(*Consumers[...]).Start.func1
```
**To Reproduce**
1. Start Tempo (SHA or version)
2. Perform Operations (Read/Write/Others)

**Expected behavior**
Running in distributed mode with multiple ingesters and distributors should not cause upstream errors of ingesters not avail.

**Environment:**
 - Infrastructure:  eks k8s 1.29
 - Deployment tool: helm distributed 2.6
 - tempo components are running on spot instances

**Additional Context**
Otel collector version - aws-otel-collector:v0.38.2

**Comment by joe-elliott** [comment]
What replication factor are you using?

**Comment by edgarkz** [comment]
@joe-elliott  We are running with default rf configuration (afaik its 3) in tempo 2.6.
we have total of 6 ingesters running.

Should i reduce rf to 1 or 2 until the new architecture is introduced?  any consequences for such an action?
https://grafana.com/docs/tempo/latest/configuration/#ingester
            replication_factor: 3

**Comment by joe-elliott** [comment]
No, leave your RF config as is. You should not be seeing this error with RF3. Is the distributor logging any errors? Can you check your ingester ring page and see if you have any unhealthy ingesters?

**Comment by edgarkz** [comment]
@joe-elliott 
log from distributor, looking at past 7 days we have dozen of failed pushes for the same reason
`
level=error ts=2025-01-02T15:31:41.029775106Z caller=rate_limited_logger.go:27 msg="pusher failed to consume trace data" err="rpc error: code = Unknown desc = Ingester is shutting down" level=error ts=2025-01-02T15:31:41.141843048Z caller=rate_limited_logger.go:27 msg="pusher failed to consume trace data" err="rpc error: code = Unknown desc = Ingester is shutting down" level=error ts=2025-01-02T15:31:41.253905116Z caller=rate_limited_logger.go:27 msg="pusher failed to consume trace data" err="rpc error: code = Unknown desc = Ingester is shutting down"
--

`
i can see that 1-3 ingestors out of 6 can shutdown due to aws spot interuptions at the same time but yet the other 3 are running fine all the time. (running on different nodes)

**Comment by joe-elliott** [comment]
In RF3 Tempo can only tolerate one ingester down at a time. If two ingesters are down it will start to refuse traces b/c it can't get write quorum. This will change with the RF1 architecture where multiple ingesters can be down but your queue will just backup.

Do you think these errors are occurring while two ingesters are unavailable? 
Are the ingesters being shut down gracefully?

**Comment by edgarkz** [comment]
@joe-elliott, regarding graceful shutdown: Yes, we have a spot interruption handler in place.

Could you please elaborate on RF3 with multiple ingestors? I was under the impression that if we run more than 3 (as we currently have 6 ingestors), the system should be able to tolerate failures as long as at least 3 remain operational out of 6.

**Comment by joe-elliott** [comment]
So in a traditional quorum based database you have to make sure that at least 1 read overlaps with 1 write to offer standard consistency guarantees. The magic formula is `w + r > RF` where w is write quorum, r is read quorum and RF is the number of replicas of the data you wrote.

With RF = 3 there are some options but we use a very standard write quorum = 2 and read quorum = 2. This means that when you write data to Tempo it attempts to write it to 3 ingesters but it will still return success if only 2 ingesters succeed. It also means that when you read a trace Tempo will return when only 2 out of 3 ingesters return results. Since 2 + 2 > 3 we are guaranteed there is overlap between the read and write and we are returning the data correctly. It's also cool that this system increases availability b/c 1 ingester can be down and we can still accept writes and return reads.

So when you write a trace to Tempo it is assigned to 3 of your 6 ingesters and if 2 of those are failing for any reason then the distributor returns failure to your OTel collector.

**Comment by edgarkz** [comment]
@joe-elliott thank you so much for the detailed explanation. It's incredibly informative and provides excellent insights into Tempo's internals. Could I kindly request that this information be added to the wiki or the Tempo ingestor page? It would greatly help operators configure and manage Tempo more effectively.

Thanks again! This issue can be closed.

**Comment by meSATYA** [comment]
@joe-elliott is the production RF1 implementation done for ingesters? We still see this issue in Tempo 2.8.1.

**Comment by edgarkz** [comment]
@meSATYA  RF1 will be introduced later this year or early next year, current tempo 2.8 version still works with RF3
